openman.tproj/openman.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 8/4/10               \" TODO: fix date
.Dt openman 1
.Os Darwin
.Sh NAME
.Nm openman
.Nd display manual pages in ManOpen.app.
.Sh SYNOPSIS
.Nm
.Op Fl kb
.Op Fl m Ar path
.Op Fl M Ar path
.Op Fl f Ar file
.Op Ar section
.Op Ar name ...
.Sh DESCRIPTION
.Nm
is a command-line utility to open Unix man pages in ManOpen.app. The
syntax is generally the same as the
.Xr man 1
command, with an additional option to directly open files.
.Pp
.Nm
will open a page in ManOpen.app for every title given.  If a section
specifier is given,
.Nm
looks in that section of the manual for the given
.Ar titles.
.Ar Section
is typically an Arabic section number (
.Dq 1
for user commands,
.Dq 8
for administrative commands, etc), but it can be a named
section as well.  If
.Ar section
is omitted,
.Nm
searches all sections of the manual, giving preference to commands
over subroutines in system libraries, and printing the first section
it finds, if any.
.Pp
If the
.Fl k
flag is specified, then
.Xr apropos 1
mode is used, with each given title
used as an Apropos lookup in ManOpen.app instead of being opened as
an individual page.
.Pp
Normally, ManOpen.app is brought forward to be the active application
when messaged by openman (meaning Terminal.app will no longer be the
active application).  If the
.Fl b
flag is specfied, ManOpen.app not be forcibly be made active (i.e.
will stay in the background).
.Pp
The man search path can be specified with the
.Fl M
or
.Fl m
flag.  The search path is a colon (':') separated list of directories
in which manual subdirectories may be found; e.g.
.Dq /usr/local/man:/usr/share/man
\&.
.Hw MANPATH
If a search path is not supplied, the value of the environment variable
.Sq MANPATH
is used for the search path.  If that isn't set, ManOpen's
man path (as specified in the application's preferences) is used.
.Pp
A file can be directly specified with the
.Fl f
flag.  If it's an nroff source file, ManOpen's
.Dq Nroff command
(as set in
the applications's preferences) will be used to process the file,
otherwise ManOpen will open it directly.  The argument can be a full or
relative path.
.Pp
If you use the tcsh shell, you can set openman's completion settings to
be similar to
.Xr man 1
's, which causes it to complete using command names.
Add the following to your ~/.cshrc or ~/.tchrc:
.Pp
.Dl complete openman 'n/-M/d/' 'p/*/c/'
.Pp
.Sh ENVIRONMENT
.Bl -tag -width "MANPATH" -indent
.It Ev MANPATH
A colon-seperated list of paths to search for man pages.
.El
.Sh SEE ALSO
ManOpen.app,
.Xr man 1 ,
.Xr apropos 1 ,
.Xr whereis 1
.Sh AUTHOR
.An Carl Lindberg Aq lindberg@clindberg.org
.Pp
Please send any bug reports/suggestions.


openman.tproj/openman.m

#import <Foundation/Foundation.h>
#import <AppKit/NSWorkspace.h>
#include <stdlib.h>


static NSString *MakeNSStringFromPath(const char *filename) NS_RETURNS_RETAINED;
NSString *MakeNSStringFromPath(const char *filename)
{
    return [[NSFileManager defaultManager] stringWithFileSystemRepresentation:filename length:strlen(filename)];
}

static NSString *MakeAbsolutePath(const char *filename) NS_RETURNS_RETAINED;
NSString *MakeAbsolutePath(const char *filename)
{
    NSString *currFile = MakeNSStringFromPath(filename);
    
    if (!currFile.absolutePath) {
        NSURL *theURL = [NSURL fileURLWithPath:currFile];
        theURL = [theURL absoluteURL];
        currFile = [theURL path];
    }
    
    return currFile;
}

static inline void usage(const char *progname)
{
    fprintf(stderr, "%s: [-bk] [-M path] [-f file] [section] [name ...]\n", progname);
}

int main (int argc, char * const *argv)
{
    @autoreleasepool {
        NSString          *manPath = nil;
        NSString          *section = nil;
        NSMutableArray    *files = [NSMutableArray array];
        BOOL              aproposMode = NO;
        BOOL              forceToFront = YES;
        NSInteger         argIndex;
        char              c;
        NSInteger         maxConnectTries = 8;
        NSInteger         connectCount = 0;
        CFErrorRef        lsError = NULL;
        NSArray           *manOpenURLs;
        if (@available(macOS 12.0, *)) {
            manOpenURLs = [[NSWorkspace sharedWorkspace] URLsForApplicationsWithBundleIdentifier:@"org.clindberg.ManOpen"];
        } else {
            manOpenURLs = CFBridgingRelease(LSCopyApplicationURLsForBundleIdentifier(CFSTR("org.clindberg.ManOpen"), &lsError));
        }
        NSMutableDictionary *distributedDictionary = [[NSMutableDictionary alloc] init];
        NSMutableArray    *namesAndSections = [[NSMutableArray alloc] init];
        CFMessagePortRef  remotePort;
        SInt32            status;
        
        while ((c = getopt(argc,argv,"hbm:M:f:kaCcw")) != EOF)
        {
            switch(c)
            {
                case 'm':
                case 'M':
                    manPath = MakeNSStringFromPath(optarg);
                    break;
                case 'f':
                    [files addObject:MakeAbsolutePath(optarg)];
                    break;
                case 'b':
                    forceToFront = NO;
                    break;
                case 'k':
                    aproposMode = YES;
                    break;
                case 'a':
                case 'C':
                case 'c':
                case 'w':
                    // MacOS X man(1) options; no-op here.
                    break;
                case 'h':
                case '?':
                default:
                    usage(argv[0]);
                    return 0;
            }
        }
        
        if (optind >= argc && [files count] <= 0)
        {
            usage(argv[0]);
            //exit(0);
            return 0;
        }
        
        if (optind < argc && !aproposMode)
        {
            NSString *tmp = @(argv[optind]);
            
            if (isdigit(argv[optind][0])          ||
                /* These are configurable in /etc/man.conf; these are just the default strings.  Hm, they are invalid as of Panther. */
                [tmp isEqualToString:@"system"]   ||
                [tmp isEqualToString:@"commands"] ||
                [tmp isEqualToString:@"syscalls"] ||
                [tmp isEqualToString:@"libc"]     ||
                [tmp isEqualToString:@"special"]  ||
                [tmp isEqualToString:@"files"]    ||
                [tmp isEqualToString:@"games"]    ||
                [tmp isEqualToString:@"miscellaneous"] ||
                [tmp isEqualToString:@"misc"]     ||
                [tmp isEqualToString:@"admin"]    ||
                [tmp isEqualToString:@"n"]        || // Tcl pages on >= Panther
                [tmp isEqualToString:@"local"])
            {
                section = tmp;
                optind++;
            }
        }
        
        if (optind >= argc)
        {
            if ([section length] > 0)
            {
                /* MacOS X assumes it's a man page name */
                section = nil;
                optind--;
            }
            
            if (optind >= argc && [files count] <= 0)
            {
                return 0;
            }
        }
        
        // Use Launch Services to find ManOpen.
        if (!manOpenURLs)
        {
            fprintf(stderr, "Cannot locate ManOpen.\n");
            return 1;
        }
        else if (![manOpenURLs.firstObject isKindOfClass:[NSURL class]])
        {
            fprintf(stderr, "Received an unknown object type from Launch Services.\n");
            return 1;
        }
        
        // Use NSWorkspace to launch ManOpen.
        if (@available(macOS 10.15, *)) {
            NSWorkspaceOpenConfiguration *configuration = [[NSWorkspaceOpenConfiguration alloc] init];
            dispatch_semaphore_t launchLock = dispatch_semaphore_create(0);
            
            configuration.activates = forceToFront;
            [[NSWorkspace sharedWorkspace] openApplicationAtURL:manOpenURLs.firstObject configuration:configuration completionHandler:^(NSRunningApplication * _Nullable app, NSError * _Nullable error) {
                if (!app)
                {
                    fprintf(stderr, "Could not launch ManOpen\n");
                    exit(1);
                }
                dispatch_semaphore_signal(launchLock);
            }];
            dispatch_semaphore_wait(launchLock, dispatch_time(DISPATCH_TIME_NOW, 10000000000)); // wait until the open event has gone through
        } else {
            NSError *launchErr = nil;
            
            if (![[NSWorkspace sharedWorkspace] launchApplicationAtURL:manOpenURLs.firstObject options:forceToFront ? 0UL : NSWorkspaceLaunchWithoutActivation configuration:@{} error:&launchErr])
            {
                fprintf(stderr, "Could not launch ManOpen\n");
                exit(1);
            }
        }
        
        // Use a Mach port to open a connection; keep trying until one connects.
        do {
            remotePort = CFMessagePortCreateRemote(NULL, CFSTR("8D98N325TG.org.clindberg.ManOpen.MachIPC"));
            if (!remotePort)
                sleep(1);
        } while (remotePort == nil && connectCount++ < maxConnectTries);
        
        if (remotePort == nil)
        {
            fprintf(stderr,"Could not open connection to ManOpen\n");
            return 1;
        }
        
        if (files.count) {
            distributedDictionary[@"Files"] = files;
        }
        
        if (manPath == nil && getenv("MANPATH") != NULL)
            manPath = MakeNSStringFromPath(getenv("MANPATH"));
        
        for (argIndex = optind; argIndex < argc; argIndex++)
        {
            NSString *currFile = MakeNSStringFromPath(argv[argIndex]);
            NSDictionary *nameAndSection;
            
            if (section)
                nameAndSection = @{@"Name": currFile, @"Section": section};
            else
                nameAndSection = @{@"Name": currFile};
            [namesAndSections addObject:nameAndSection];
        }
        if (aproposMode) {
            distributedDictionary[@"Apropos"] = @YES;
        }
        if (manPath) {
            distributedDictionary[@"ManPath"] = manPath;
        }
        distributedDictionary[@"NamesAndSections"] = namesAndSections;
        
        // Once we've got all our data together, send the message to the app. The message ID below is intentional, because I feel like I went mental writing this (it was my third attempt at replacing the original DO code).
        status = CFMessagePortSendRequest(remotePort, 5150, (CFDataRef)[NSKeyedArchiver archivedDataWithRootObject:distributedDictionary requiringSecureCoding:YES error:NULL], 10.0, 10.0, NULL, NULL);
        return status == kCFMessagePortSuccess ? EXIT_SUCCESS : EXIT_FAILURE;
    }
}
